src/samples/audio/smart_amp_test_ipc4.c

```diff
@@ -37,57 +55,50 @@ typedef void (*smart_amp_proc)(int8_t const *src_ptr,
 struct smart_amp_data {
 	struct sof_smart_amp_ipc4_config ipc4_cfg;
 	struct sof_smart_amp_config config;
-	struct comp_data_blob_handler *model_handler;
-	void *data_blob;
-	size_t data_blob_size;
 	smart_amp_proc process;
 	uint32_t out_channels;
 };
 
+#ifdef __MODULE_BUILD__
+static struct smart_amp_data smart_amp_priv;
+#endif
+
 static int smart_amp_init(struct processing_module *mod)
 {
 	struct smart_amp_data *sad;
-	struct comp_dev *dev = mod->dev;
 	struct module_data *mod_data = &mod->priv;
-	const size_t in_size = sizeof(struct ipc4_input_pin_format) * SMART_AMP_NUM_IN_PINS;
-	const size_t out_size = sizeof(struct ipc4_output_pin_format) * SMART_AMP_NUM_OUT_PINS;
 	int ret;
 	const struct ipc4_base_module_extended_cfg *base_cfg = mod_data->cfg.init_data;
 
-	comp_dbg(dev, "smart_amp_init()");
+	LOG_DBG("smart_amp_init()");
+
+#ifndef __MODULE_BUILD__
 	sad = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*sad));
 	if (!sad)
 		return -ENOMEM;
-
+#else
+	sad = &smart_amp_priv;
+#endif
 	mod_data->private = sad;
 
-	/* component model data handler */
-	sad->model_handler = comp_data_blob_handler_new(dev);
-	if (!sad->model_handler) {
-		ret = -ENOMEM;
-		goto sad_fail;
-	}
-
 	if (base_cfg->base_cfg_ext.nb_input_pins != SMART_AMP_NUM_IN_PINS ||
 	    base_cfg->base_cfg_ext.nb_output_pins != SMART_AMP_NUM_OUT_PINS) {
-		comp_err(dev, "smart_amp_init(): Invalid pin configuration");
 		ret = -EINVAL;
 		goto sad_fail;
 	}
 
 	/* Copy the pin formats */
-	memcpy_s(sad->ipc4_cfg.input_pins, in_size,
-		 base_cfg->base_cfg_ext.pin_formats, in_size);
-	memcpy_s(&sad->ipc4_cfg.output_pin, out_size,
-		 &base_cfg->base_cfg_ext.pin_formats[in_size], out_size);
-
-	mod->max_sources = SMART_AMP_NUM_IN_PINS;
+	sad->ipc4_cfg.input_pins[0] =
+		*((struct ipc4_input_pin_format *)base_cfg->base_cfg_ext.pin_formats);
+	sad->ipc4_cfg.input_pins[1] =
+		*((struct ipc4_input_pin_format *)base_cfg->base_cfg_ext.pin_formats + 1);
+	sad->ipc4_cfg.output_pin =
+		*(struct ipc4_output_pin_format *)(base_cfg->base_cfg_ext.pin_formats
+			+ sizeof(sad->ipc4_cfg.input_pins));
 
 	return 0;
 
 sad_fail:
-	comp_data_blob_handler_free(sad->model_handler);
-	rfree(sad);
```

Comment:
you still need `rfree()` for monolithic builds, don't you?

Reply:
sure, thanks